etc/library.sh

```diff
@@ -114,51 +119,151 @@ function run_app_unsafe()
+      jq -e '.tmux' <<<"$cfg" > /dev/null 2>&1
+      use_tmux="$?"
+      len="$(jq '.params | length' <<<"$cfg")"
+      for (( i = 0 ; i < len ; i++ )); do
+        local var val
+        var="$(jq -r ".params[$i].id"    <<<"$cfg")"
+        val="$(jq -r ".params[$i].value" <<<"$cfg")"
+        eval "export $var=$val"
+      done
+    }
+ 
+    echo "$ncp_app" > "$LOCK_FILE"
+    if which tmux > /dev/null && [[ $use_tmux == 0 ]]
+    then
+      echo "Running $ncp_app in tmux..." | tee -a $log
+      # Run app in tmux
+      local tmux_log_file tmux_status_file LIBPATH
+      tmux_log_file="/var/log/ncp/tmux.${ncp_app}.log"
+      tmux_status_file="/var/log/ncp/tmux.${ncp_app}.status"
+      LIBPATH="$(dirname $CFGDIR)/library.sh"
+      
+      # Reset tmux output
+      echo "[ $ncp_app ]" | tee -a $log
+      echo "[ $ncp_app ]" > "$tmux_log_file"
+      echo "" > "$tmux_status_file"
+      chmod 640           "$tmux_log_file" "$tmux_status_file"
+      chown root:www-data "$tmux_log_file" "$tmux_status_file"
+
+      tmux new-session -d -s "$ncp_app" "bash -c '(
+        trap \"echo \\\$? > $tmux_status_file && rm $LOCK_FILE\" 1 2 3 4 6 9 11 15 19 29
+        source \"$LIBPATH\"
+        source \"$script\"
+        configure 2>&1 | tee -a $log
+        echo \"\${PIPESTATUS[0]}\" > $tmux_status_file
+        rm $LOCK_FILE
+      )' 2>&1 | tee -a $tmux_log_file"
+
+      attach_to_app "$ncp_app"
+      exit
+
+    else
+      trap "rm '$LOCK_FILE'" 0 1 2 3 4 6 11 15 19 29
+      echo "[ $ncp_app ]" | tee -a $log
+      echo "Running $ncp_app directly..." | tee -a $log
+      # read script
+      # shellcheck source=/dev/null
+      source "$script"
+      # run
+      configure 2>&1 | tee -a $log
+      local ret="${PIPESTATUS[0]}"
+      exit "$ret"
+    fi
+  )
+  ret="$?"
   echo "" >> $log
 
   return "$ret"
 }
 
+function attach_to_app()
+{
+  local tmux_log_file tmux_status_file
+  tmux_log_file="/var/log/ncp/tmux.${ncp_app}.log"
+  tmux_status_file="/var/log/ncp/tmux.${ncp_app}.status"
+
+  if [[ "$ATTACH_NO_FOLLOW" == "1" ]]
+  then
+    cat "$tmux_log_file"
+    return 0
+  else
+    (while tmux has-session -t="$ncp_app" > /dev/null 2>&1 
+    do
+      sleep 1
+    done) &
+
+    # Follow log file until tmux session has terminated
+    tail --lines=+0 -f "$tmux_log_file" --pid="$!"
+  fi
+
+  # Read return value from tmux log file
+  ret="$(tail -n 1 "$tmux_status_file")"
+  #rm "$tmux_log_file"
+  #rm "$tmux_status_file"
+
+  [[ $ret =~ ^[0-9]+$ ]] && return $ret
+  return 1
+}
+
 function is_active_app()
 {
   local ncp_app=$1
   local bin_dir=${2:-.}
   local script="$bin_dir/$ncp_app.sh"
   local cfg_file="$CFGDIR/$ncp_app.cfg"
 
-  [[ -f "$script" ]] || local script="$(find "$BINDIR" -name $ncp_app.sh)"
+  [[ -f "$script" ]] || script="$(find "$BINDIR" -name $ncp_app.sh)"
   [[ -f "$script" ]] || { echo "file $script not found"; return 1; }
 
   # function
   unset is_active
+  # shellcheck source=/dev/null
```

Comment:
what does this do?

Reply:
It disables a shellcheck warning about the following line where the sourced file is determined dynamically and can thus not be accounted for by shellcheck.

Reply:
https://github.com/koalaman/shellcheck/wiki/SC1090